curvefs/src/metaserver/dentry_storage.cpp

```diff
@@ -235,92 +264,344 @@ MetaStatusCode DentryStorage::Find(const Dentry& in,
+    if (s.IsNotFound()) {
+        return MetaStatusCode::NOT_FOUND;
+    }
+    return MetaStatusCode::STORAGE_INTERNAL_ERROR;
+}
+
+storage::Status DentryStorage::SetAppliedIndex(
+    storage::StorageTransaction* transaction, int64_t index) {
+    common::AppliedIndex val;
+    val.set_index(index);
+    return transaction->SSet(table4AppliedIndex_, kDentryAppliedKey, val);
+}
+
+storage::Status DentryStorage::DelAppliedIndex(
+    storage::StorageTransaction* transaction) {
+    return transaction->SDel(table4AppliedIndex_, kDentryAppliedKey);
+}
+
+storage::Status DentryStorage::SetHandleTxIndex(
+    storage::StorageTransaction* transaction, int64_t index) {
+    common::AppliedIndex val;
+    val.set_index(index);
+    return transaction->SSet(table4AppliedIndex_, kHandleTxKey, val);
+}
+
+storage::Status DentryStorage::DelHandleTxIndex(
+    storage::StorageTransaction* transaction) {
+    return transaction->SDel(table4AppliedIndex_, kHandleTxKey);
+}
+
+MetaStatusCode DentryStorage::GetPendingTx(
+    metaserver::TransactionRequest* request) {
+    auto s = kvStorage_->SGet(table4Transaction_, kPendingTxKey, request);
+    if (s.ok()) {
+        return MetaStatusCode::OK;
+    }
+    if (s.IsNotFound()) {
+        return MetaStatusCode::NOT_FOUND;
+    }
+    return MetaStatusCode::STORAGE_INTERNAL_ERROR;
+}
+
+storage::Status DentryStorage::SetPendingTx(
+    storage::StorageTransaction* txn,
+    const metaserver::TransactionRequest& request) {
+    return txn->SSet(table4Transaction_, kPendingTxKey, request);
+}
+
+storage::Status DentryStorage::DelPendingTx(storage::StorageTransaction* txn) {
+    return txn->SDel(table4Transaction_, kPendingTxKey);
+}
+
+storage::Status DentryStorage::ClearPendingTx(
+    storage::StorageTransaction* txn) {
+    metaserver::TransactionRequest request;
+    request.set_type(metaserver::TransactionRequest::None);
+    request.set_rawpayload("");
+    return txn->SSet(table4Transaction_, "transaction", request);
+}
+
+storage::Status DentryStorage::SetDentryCount(storage::StorageTransaction* txn,
+                                              uint64_t count) {
+    common::ItemCount val;
+    val.set_count(count);
+    return txn->SSet(table4DentryCount_, kDentryCountKey, val);
+}
+
+storage::Status DentryStorage::DelDentryCount(
+    storage::StorageTransaction* txn) {
+    return txn->SDel(table4DentryCount_, kDentryCountKey);
+}
+
+storage::Status DentryStorage::GetDentryCount(uint64_t* count) {
+    common::ItemCount val;
+    auto s = kvStorage_->SGet(table4DentryCount_, kDentryCountKey, &val);
+    if (s.ok()) {
+        *count = val.count();
+    }
+    return s;
+}
+
+storage::Status DentryStorage::GetHandleTxIndex(int64_t* index) {
+    common::AppliedIndex val;
+    auto s = kvStorage_->SGet(table4AppliedIndex_, kHandleTxKey, &val);
+    if (s.ok()) {
+        *index = val.index();
+    }
+    return s;
+}
+
+MetaStatusCode DentryStorage::Insert(const Dentry& dentry, int64_t logIndex) {
```

Comment:
I think the following functions involving transactions are a bit complex and there are a lot of repetitive logic，it is a bit difficult for most people to understand. can we try to simplify it, like `Insert`, `Delete` and etc.

Reply:
> I think the following functions which related to transaction are a bit complex and there are a lot of repetitive logic, it is a bit difficult for most people to understand. can we try to simplify it? like `Insert`, `Delete` and etc. The transaction is a 2PC implemation which include `Prepare` and `Commit` phase, i think our implemation should make this point very clear to the reader :)

Which function related to transaction do you consider to be complex?